File: 21.py
"""
Let d(n) be defined as the sum of proper divisors of n (numbers less than n 
which divide evenly into n). If d(a) = b and d(b) = a, where a ≠ b, then a 
and b are an amicable pair and each of a and b are called amicable numbers.

For example, the proper divisors of 220 are 1, 2, 4, 5, 10, 11, 20, 22, 44,
55 and 110; therefore d(220) = 284. The proper divisors of 284 are 1, 2, 4,
71 and 142; so d(284) = 220.

Evaluate the sum of all the amicable numbers under 10000.
"""


def amicable(n):
    i = 1
    divisors = set()
    while i <= n / 2:
        if n % i == 0:
            if i in divisors:
                break
            else:
                divisors.add(i)
                if i > 1:
                    divisors.add(int(n / i))
        i += 1

    return sum(divisors)


amicable_nums = set()
for i in range(1, 10000):
    a, b = i, amicable(i)
    db = amicable(b)

    if (a == db) & (a != b):
        amicable_nums.add(a)
        amicable_nums.add(b)

print(sum(amicable_nums))

File: 21.cpp
#include <iostream>
#include <numeric>
#include <vector>

using std::accumulate;
using std::cout;
using std::endl;
using std::vector;

int amicable(int n) {
  int i = 1;
  vector<int> divisors;

  while (i <= n / 2) {
    if (n % i == 0) {
      divisors.push_back(i);
    }
    ++i;
  }

  int sum = std::accumulate(divisors.begin(), divisors.end(), 0);

  return sum;
}

int main() {
  int amicable_sum;

  for (int i = 220; i < 10000; ++i) {
    int b = amicable(i);
    int db = amicable(b);

    if (i == db && i != b) {
      amicable_sum = amicable_sum + i;
    }
  }

  cout << amicable_sum << endl;
  return 0;
}